_shared/local_extensions.py

```diff
@@ -214,6 +215,14 @@ def random_port_number(self):
 
         return self._random_port_numbers.pop()
 
+    @pass_context
+    def has_services(self, context):
+        """Return True if the project has services."""
+        cookiecutter = context["cookiecutter"]
+        return cookiecutter.get("db") == "yes" or self.include_exists(
+            context, "docker-compose/services.yml"
+        )
```

Comment:
This boolean expression was going to get repeated again and again in the templates. We have no way to share code between template files except with the global template context functions in this `local_extensions.py` file. The repetition was particularly egregious and would have been a maintenance issue so I thought it was worth adding this boolean function to deduplicate it

Reply:
> Seems sensible. I do wonder if the "no way to share code" is something we should investigate changing. It seems like quite a speed bump for nice maintainable templates.

I've looked into this one previously. It's a limitation of cookiecutter and not something we can work around, at least not without the cure being worse than the disease. I think if we want to change it we should send a PR to cookiecutter and get them to merge it and release a new version. But it's probably not worth the effort: our templates aren't actually suffering that much from this.

Reply:
I don't think it makes sense to have a `services=yes` configuration option because it's not sensible for a project to just say "I want services", the project has to actually say _which_ service(s) it wants: Postgres, Elasticsearch, etc. If all the project has said is "I want services" then the cookiecutter cannot generate a valid `docker-compose.yml` file (there's nothing to put in the `services` section), it cannot generate a valid `services` segment within the `ci.yml` file, etc.

So the way it works is that instead of `services=yes` we have options for specific services, currently this is only `db=yes`. So the project says `db=yes` and the cookiecutter is able to output a valid `docker-compose.yml` file with a `postgres` service in it, a follow `services` section within the `ci.yml` file with a `postgres` service in it, etc.